Sofa/framework/Core/src/sofa/core/behavior/MatrixAPICompatibility.h

```diff
@@ -0,0 +1,214 @@
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+
+#include <sofa/linearalgebra/BaseMatrix.h>
+#include <sofa/core/MatrixAccumulator.h>
+#include <sofa/core/behavior/DefaultMultiMatrixAccessor.h>
+#include <sofa/core/behavior/BaseProjectiveConstraintSet.h>
+
+namespace sofa::core::behavior
+{
+
+/// This class exists only for compatibility reasons. To be removed once the deprecated API
+/// addKToMatrix and addMToMatrix is removed
+class MatrixAccessorCompat : public sofa::core::behavior::DefaultMultiMatrixAccessor
+{
+public:
+    MatrixRef getMatrix(const sofa::core::behavior::BaseMechanicalState* mstate) const override
+    {
+        MatrixRef r;
+        const auto& matrix = matrixMap.at(sofa::type::fixed_array<const sofa::core::behavior::BaseMechanicalState*, 2>{mstate, mstate});
+        r.matrix = matrix.get();
+        r.offset = 0;
+        return r;
+    }
+
+
+    InteractionMatrixRef getMatrix(const sofa::core::behavior::BaseMechanicalState* mstate1, const sofa::core::behavior::BaseMechanicalState* mstate2) const override
+    {
+        InteractionMatrixRef r;
+        const auto& matrix = matrixMap.at(sofa::type::fixed_array<const sofa::core::behavior::BaseMechanicalState*, 2>{mstate1, mstate2});
+        r.matrix = matrix.get();
+        r.offRow = 0;
+        r.offCol = 0;
+        return r;
+    }
+
+    void setMatrix(sofa::core::behavior::BaseMechanicalState* mstate1, sofa::core::behavior::BaseMechanicalState* mstate2, std::shared_ptr<linearalgebra::BaseMatrix> matrix)
+    {
+        matrixMap[sofa::type::fixed_array<const sofa::core::behavior::BaseMechanicalState*, 2>{mstate1, mstate2}] = matrix;
+    }
+
+    std::map<
+        sofa::type::fixed_array<const sofa::core::behavior::BaseMechanicalState*, 2>, std::shared_ptr<linearalgebra::BaseMatrix> > matrixMap;
+};
+
+/// A fake BaseMatrix redirecting its add methods to the MatrixAccumulator API
+/// This class exists only for compatibility reasons. To be removed once the deprecated API
+/// addKToMatrix, addBToMatrix and addMToMatrix is removed
+template<matrixaccumulator::Contribution c>
+class AddToMatrixCompatMatrix : public sofa::linearalgebra::BaseMatrix
+{
+public:
+    static constexpr const char* compatibilityMessage = "This message appears only for compatibility"
+        " of the deprecated API addKToMatrix, addBToMatrix and addMToMatrix. Update your code with "
+        "the new API buildStiffnessMatrix, buildDampingMatrix or buildMassMatrix to remove this warning. ";
+
+    ~AddToMatrixCompatMatrix() override = default;
+    Index rowSize() const override
+    {
+        msg_error(component) << compatibilityMessage << "rowSize is not a supported operation in the compatibility";
+        return {};
+    }
```

Comment:
(and all the other methods)
The messages are marked "error" (i.e the user must not do that), is it really an error ? or just a (strong) warning ?
And I think this will spam a lot of lines in the debug logs (console, etc) but maybe this is your intention.

Reply:
Actually, the mentioned methods are not supposed to be called. So no endless logs in theory.
This "matrix" is used only for the compatibility with the old API (i.e. `addKToMatrix` etc). Only the `add` methods are usually called in `addKToMatrix`. The goal is obviously to implement `buildStiffnessMatrix` to stop using the compatibility layer.